### ISSUE TITLE
Type inference gets confused by method call in closure

### ISSUE THREAD

**Issue by JakobDegen** [rust-lang/rust#92424]
On both `rustc 1.59.0-nightly (83b15bfe1 2021-12-28)` and `rustc 1.57.0 (f1edd0429 2021-11-29)`, this code:

```rust
fn main() {
    let _ = |s| (s.wrapping_add(1), u8::wrapping_add(s, 1));
}
```

Errors:

```
error[E0282]: type annotations needed
 --> src/main.rs:2:14
  |
2 |     let _ = |s| (s.wrapping_add(1), u8::wrapping_add(s, 1));
  |              ^ consider giving this closure parameter a type
  |
  = note: type must be known at this point

For more information about this error, try `rustc --explain E0282`.
```

Annotating the parameter with `: u8` fixes the error. Annotating the output type does not. Replacing the body with:

```rust
let _ = |s| (1, u8::wrapping_add(s, 1));
```

also does not produce the error.

I may not understand type inference well enough, but this feels like a bug. If the type of `s` can be determined from the function call in the second example, I don't see why the addition of a method call on `s` should cause type inference to fail.

**Comment by JakobDegen** [comment]
Ah, this is fixed by switching the order of the expressions in the tuple. This is apparently a known restriction, so I'm going to close this for now. If anyone finds something concrete here to work on, feel free to re-open
